skimage/transform/_thin_plate_splines.py

```diff
@@ -0,0 +1,302 @@
+import numpy as np
+import scipy as sp
+
+
+def tps_warp(
+    image,
+    src,
+    dst,
+    output_region,
+    interpolation_order=1,
+    grid_scaling=None
+):
+    """Return an array of warped images.
+
+    Define a thin-plate-spline warping transform that warps from the
+    src to the dst, and then warp the given images by
+    that transform.
+
+    Parameters
+    ----------
+    src: (N, 2) array_like
+        Source image coordinates.
+    dst: (N, 2) array_like
+        Destination image coordinates.
+    image: ndarray
+        Input image.
+    output_region: (1, 4) array
+        The (xmin, ymin, xmax, ymax) region of the output
+        image that should be produced. (Note: The region is inclusive, i.e.
+        xmin <= x <= xmax)
+    interpolation_order: int, optional
+        If value is 1, then use linear interpolation else use
+        nearest-neighbor interpolation.
+    grid_scaling: int, optional
+        If grid_scaling is greater than 1, say x, then the transform is
+        defined on a grid x times smaller than the output image region.
+        Then the transform is bilinearly interpolated to the larger region.
+        This is fairly accurate for values up to 10 or so.
+
+    Returns
+    -------
+    warped: array_like
+        Array of warped images.
+
+    Examples
+    --------
+    Produce a warped image rotated by 90 degrees counter-clockwise:
+
+    >>> import skimage as ski
+    >>> image = ski.data.astronaut()
+    >>> astronaut = ski.color.rgb2gray(image)
+    >>> src = np.array([[0, 0], [0, 500], [500, 500],[500, 0]])
+    >>> dst = np.array([[500, 0], [0, 0], [0, 500],[500, 500]])
+    >>> output_region = (0, 0, astronaut.shape[1], astronaut.shape[0])
+    >>> warped_image = ski.transform.tps_warp(image, src, dst,
+    ...     output_region=output_region)
+    References
+    ----------
+    .. [1] Bookstein, Fred L. "Principal warps: Thin-plate splines and the
+    decomposition of deformations." IEEE Transactions on pattern analysis and
+    machine intelligence 11.6 (1989): 567–585.
+
+    """
+    if image.size == 0:
+        raise ValueError("Cannot warp empty image with dimensions", image.shape)
+
+    if image.ndim != 2:
+        raise ValueError("Only 2-D images (grayscale or color) are supported")
+
+    transform = _make_inverse_warp(src, dst,
+                                   output_region, grid_scaling)
+    return sp.ndimage.map_coordinates(np.asarray(image), transform, order=interpolation_order)
```

Comment:
I'd move the `np.asarray(image)` bit to the start of the function before you call `image.size`, etc. Otherwise, if something other than an array is passed in, you get less useful "AttributeError: no attribute 'size'" errors..